### PR TITLE
Introducing documentSelector for Document Targeting

### LINKED ISSUE #179: documentIndex is brittle
Our tests include template files which produce yaml files with many documents, in which the ordering might often change.

Is there a way to select a document with a yaml-path or something like that?

For example, we now have something like this:

```yaml
suite: prometheus
templates:
  - addons/default-values.yaml
tests:
  - it: X
    set:
      prometheusOperator.enabled: true
    values: [ generic-x.yaml ]
    documentIndex: 11 # Prometheus. Index-based is pretty brittle if the ordering ever changes
    asserts:
      - isKind:
          of: ConfigMap
      - matchSnapshot: {}
  - it: non-x
    values: [ generic-y.yaml ]
    documentIndex: 8 # Prometheus. Index-based is pretty brittle if the ordering ever changes
    asserts:
      - isKind:
          of: ConfigMap
      - matchSnapshot: {}

```

This would immediately break if our addons/default-values template will output extra ConfigMaps due to our base values being changed.

I'd expect something like 

```yaml
suite: prometheus
templates:
  - addons/default-values.yaml
tests:
  - it: X
    set:
      prometheusOperator.enabled: true
    values: [ generic-x.yaml ]
    documentSelector:
      - hasName: prometheus-operator-default-values
      - path: something.else
           is: "toCheck"
    asserts:
      - isKind:
          of: ConfigMap
      - matchSnapshot: {}
...

```

If there isn't an alternative to selecting a document other than its numerical index, are the maintainers open to have this implemented?

### PR DESCRIPTION
**Pull request by jonathon-mcnabb**
# Description

In response to the challenges faced with the index-based selection of documents within multi-document YAML templates, we introduce the documentSelector feature. This provides a more robust mechanism to select specific documents without relying on a brittle index.

Context: Our tests often include template files which produce YAML files containing multiple documents. The ordering of these documents might often change, making the index-based selection a fragile method. Our community voiced concerns about this, suggesting a yaml-path based selection mechanism. The presented PR is a direct answer to this request.

Fixes #179 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by jonathon-mcnabb** [comment]
@quintush  Please let me know if there are any style guidelines, test, or anything else that needs to be done to complete this PR.

**Comment by quintush** [comment]
Hello @jonathon-mcnabb,

Thanks for the great addon.
The solution looks like the way I would go to.

Regarding code guidelines, it is sufficient to run gofmt to format the code and make sure naming convention is readable.
Will come back with some feedback pointers.

Greetings,
@quintush

**Comment by hansbogert** [comment]
As an invested party, can I help in any way getting this merged/released?

**Comment by hansbogert** [comment]
@jonathon-mcnabb  If you don't have the time can I add on your work my proposal to make the the documentSelector a list?

File: pkg/unittest/valueutils/documentselector.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+		for idx, doc := range fileManifests {
+			manifestValues, err := GetValueOfSetPath(doc, selector.Path)
+			if err != nil {
+				continue
+			}
+
+			for _, manifestValue := range manifestValues {
+				if reflect.DeepEqual(selector.Value, manifestValue) {
+					return idx, nil
+				}
```
Comment (hansbogert) [inline pkg/unittest/valueutils/documentselector.go:27]: This returns the first time it matches, but if you have multiple manifests in `manifests` they could have different `idx`s.

**Review by hansbogert — COMMENTED** [review]
See 1 comment.

Also, did something go wrong with rebasing? This PR now 89 changes and  many commits.

**Comment by hansbogert** [comment]
@raxod502-plaid No, the list functionality was not implemented, afaict.

**Pull request merged**